yarn/src/main/scala/org/apache/spark/deploy/yarn/AMDelegationTokenRenewer.scala

```diff
@@ -0,0 +1,203 @@
+
+  private var loggedInViaKeytab = false
+  var driverEndPoint: RpcEndpointRef = null
+
+  private val hadoopUtil = YarnSparkHadoopUtil.get
+
+  /**
+   * Schedule a login from the keytab and principal set using the --principal and --keytab
+   * arguments to spark-submit. This login happens only when the credentials of the current user
+   * are about to expire. This method reads SPARK_PRINCIPAL and SPARK_KEYTAB from the environment
+   * to do the login. This method is a no-op in non-YARN mode.
+   */
+  private[spark] def scheduleLoginFromKeytab(): Unit = {
+    val principal = sparkConf.get("spark.yarn.principal")
+    val keytab = sparkConf.get("spark.yarn.keytab")
+
+    def getRenewalInterval: Long = {
+      import scala.concurrent.duration._
+      val credentials = UserGroupInformation.getCurrentUser.getCredentials
+      val interval = (0.75 * (hadoopUtil.getLatestTokenValidity(credentials) -
+        System.currentTimeMillis())).toLong
+//      // If only 6 hours left, then force a renewal immediately. This is to avoid tokens with
+//      // very less validity being used on AM restart.
+//      if ((interval millis).toHours <= 6) {
+//        0L
+//      } else {
+//        interval
+//      }
+      interval
+    }
+
+    def scheduleRenewal(runnable: Runnable): Unit = {
+      val renewalInterval = getRenewalInterval
+      logInfo(s"Scheduling login from keytab in $renewalInterval millis.")
+      delegationTokenRenewer.schedule(runnable, renewalInterval, TimeUnit.MILLISECONDS)
+    }
+
+    // This thread periodically runs on the driver to update the delegation tokens on HDFS.
+    val driverTokenRenewerRunnable =
+      new Runnable {
+        override def run(): Unit = {
+          try {
+            writeNewTokensToHDFS(principal, keytab)
+            cleanupOldFiles()
+          } catch {
+            case e: Exception =>
+              // Log the error and try to write new tokens back in an hour
+              logWarning("Failed to write out new credentials to HDFS, will try again in an " +
+                "hour! If this happens too often tasks will fail.", e)
+              delegationTokenRenewer.schedule(this, 1, TimeUnit.HOURS)
+              return
+          }
+          scheduleRenewal(this)
+        }
+      }
+    // Schedule update of credentials. This handles the case of updating the tokens right now
+    // as well, since the renenwal interval will be 0, and the thread will get scheduled
+    // immediately.
+    scheduleRenewal(driverTokenRenewerRunnable)
+  }
+
+  // Keeps only files that are newer than 30 days, and deletes everything else. At least 5 files
+  // are kept for safety
+  private def cleanupOldFiles(): Unit = {
+    import scala.concurrent.duration._
+    try {
+      val remoteFs = FileSystem.get(hadoopConf)
+      val credentialsPath = new Path(sparkConf.get("spark.yarn.credentials.file"))
+      val thresholdTime = System.currentTimeMillis() - (30 days).toMillis
```

Comment:
30 days, generally assume 1 file per day = 30 files.  Still seems like an awful lot around still.   I would think in reality you would only be using possibly 2, the current one and the previous one.  If its using more then that then your expiry interval is really small or something else is wrong.   Can we make a config to change this?  We can keep it undocumented for now.